### PR TITLE
A reader writer lock on the file region

### PR DESCRIPTION
**Pull request by bas524**
This PR contains implementation of RWLock on the file descriptor
It allows multiple concurrent process-readers or one exclusive process-writer on the file region

example

```cpp
// open file for reading and writing
FileStream fs(filepath, std::ios::in | std::ios::out | std::ios::binary);
// write some information
Poco::Int32 i32 = 0;
fs.seekp(0, std::ios::beg);
fs.write((const char *)&i32, sizeof(i32));
fs.flushToDisk();
//...
// other process-reader
FileStreamRWLock fLock(fs, 0, sizeof(Poco::Int32));
ScopedFStreamReadRWLock readLock(fLock);
fs.seekg(0, std::ios::beg);
fs.read((char *)&counter, sizeof(counter));

//...
// other process-writer
FileStreamRWLock fLock(fs, 0, sizeof(Poco::Int32));
ScopedFStreamWriteRWLock writeLock(fLock);
fs.seekp(0, std::ios::beg);
fs.write((char *)&counter, sizeof(counter));
fs.flushToDisk();
```

!Warning
On windows you should open file by WINAPI and then use handle with poco, because Poco::FileStream doesn't open or create file with ```FILE_SHARE_READ | FILE_SHARE_WRITE``` and ```GENERIC_READ | GENERIC_WRITE```

You can do it with this code
```cpp
HANDLE openFileWithRWAccess(const std::string& path)
{
	DWORD access = GENERIC_READ | GENERIC_WRITE;
	DWORD shareMode = FILE_SHARE_READ | FILE_SHARE_WRITE;

	HANDLE handle = CreateFileA(path.c_str(), access, shareMode, NULL, OPEN_ALWAYS, FILE_ATTRIBUTE_NORMAL, NULL);

	if (handle == INVALID_HANDLE_VALUE)
		Poco::File::handleLastError(path);

	return handle;
}
```

and then

```cpp
FileStream fs;
fs.openHandle(openFileForSahredRWAccess(fl), std::ios::in | std::ios::out | std::ios::binary);
```

### CODE REVIEW

File: Foundation/testsuite/src/FileStreamRWLockTest.cpp [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+	Poco::Path testAppPath = Poco::Path::current();
+	testAppPath.append(name + ext).makeFile();
+	return testAppPath.toString();
+}
+
+#if POCO_OS == POCO_OS_WINDOWS_NT
+
+#include "Poco/File_WIN32U.h"
+
+HANDLE createFileForSahredRWAccess(const std::string &path)
```
Comment (aleks-f) [inline Foundation/testsuite/src/FileStreamRWLockTest.cpp:57]: spelling

File: Foundation/src/FileStreamRWLock.cpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+
+FileStreamRWLock::FileStreamRWLock(const FileStream &fs, Poco::UInt64 offset, Poco::UInt64 size) :
+	FileStreamRWLockImpl(fs.nativeHandle(), offset, size)
+{
+}
+
+
+FileStreamRWLock::~FileStreamRWLock()
+{
```
Comment (aleks-f) [inline Foundation/src/FileStreamRWLock.cpp:35]: there should be an `atomic_bool _locked` flag that unlocks the file if it is still locked on destruction

**Review by aleks-f — CHANGES_REQUESTED** [review]

File: Foundation/testsuite/src/FileStreamRWLockTest.cpp (hunk 1 above)
Comment (bas524) [reply Foundation/testsuite/src/FileStreamRWLockTest.cpp]: fixed (I hope)

File: Foundation/src/FileStreamRWLock.cpp (hunk 2 above)
Comment (bas524) [reply Foundation/src/FileStreamRWLock.cpp:35]: done

**Comment by aleks-f** [comment]
@bas524 wasn't triggering al CI, moved to #4734

**Pull request closed without merging**